how-it-works/3_chainevolution/02_fault-tolerance.subpage.md

```diff
@@ -7,7 +7,41 @@ slug: fault-tolerance
 
 # Fault tolerance
 
-Since the Internet Computer is fault tolerant, the network will make progress even if some nodes fail or misbehave. If a node has been offline for a while or is being added to a subnet, efficient mechanisms are in place to ensure that such a node is able to process messages quickly. This involves learning which other nodes to connect to, gathering the necessary key material and information to participate in the consensus algorithm, and obtaining a recent checkpoint of the state and the queues of the canister smart contracts hosted on the subnet.
+
+In today’s world, softwares and networks crash surprising often. This could be because of software bugs, improper error handling, power failure, etc. The Internet Computer is designed to be fault tolerant. In other words, the network will make progress even if some nodes fail or misbehave. 
+
+## Handling Node Failures
+The nodes in the Internet Computer are divided into subnets. The nodes in a subnet together maintain a blockchain. In each round, a block is proposed by the consensus layer, and the messages in the block are processed later by the execution layer. The proposed block and the resulting state needs to be signed by at least 2/3rd of the nodes in the subnet in order to make progress. As long as less than 1/3rd of the nodes in a subnet fail or misbehave, the subnet will continue making progress. 
+
+Let us consider a scenario where less than 1/3rd of the nodes in a subnet fail while rest of the subnet continue to make progress. We will now describe how a failed node can recover automatically and catchup with the other well-functioning nodes. A newly joined node also uses the same process to catchup with the existing nodes in the subnet. 
+
+Here’s one natural solution. A failed or newly joined node could download all the consensus blocks it missed from its peers, and process each block one by one. Unfortunately, new nodes will take a long time to catchup if they have to process all the blocks from the genesis. Another solution is to let the failed or newly joined node to directly sync the latest state from its peers. However, the peers are continuously updating their state as they process blocks. Syncing the latest state while the peers are updating it could lead to inconsistencies. 
```

Comment:
```suggestion
Here’s one natural solution. A failed or newly joined node could download all the consensus blocks it missed from its peers, and process each block one by one. Unfortunately, new nodes will take a long time to catch up if they have to process all the blocks from the genesis. Another solution is to let the failed or newly joined node to directly sync the latest state from its peers. However, the peers are continuously updating their state as they process blocks. Syncing the latest state while the peers are updating it could lead to inconsistencies. 
```